backend/app/controllers/campaigns_controller.rb

```diff
@@ -48,9 +63,92 @@ def set_campaign
   end
 
   def campaign_params
-    permitted_params = [:name, :description, :promised_amount, :start, :end, :primary_donor_id,
-                        :coupon_denomination, { charity_ids: [] }]
+    top_level_params = %i[charities primary_donor]
+    params.require(top_level_params)
+
+    campaign_params = %i[name description promised_amount start end coupon_denomination]
+    params.require(:campaign).permit(campaign_params)
+  end
+
+  def set_donor
+    primary_donor_params = params[:primary_donor]
+
+    @campaign.primary_donor = PrimaryDonor.find_or_initialize_by(email: primary_donor_params[:email]) do |new_donor|
+      new_donor.name = primary_donor_params[:name]
+    end
+  end
+
+  def create_campaign_charities
+    charity_params = params[:charities]
+
+    @campaign.campaign_charities = charity_params.map do |campaign_charity_param|
+      CampaignCharity.new(charity_id: campaign_charity_param[:charity][:id],
+                          giving_sg_url: campaign_charity_param[:giving_sg_url])
+    end
+  end
+
+  def create_or_update_campaign_charities
+    charity_params = params[:charities]
+
+    existing_ids = charity_params.pluck(:id).compact
+    campaign_charities = CampaignCharity.find(existing_ids)
+
+    campaign_charities_to_add = charity_params.select { |campaign_charity| campaign_charity[:id].nil? }
+
+    campaign_charities_to_add.each do |campaign_charity|
+      campaign_charities << CampaignCharity.new(charity_id: campaign_charity[:charity][:id],
+                                                giving_sg_url: campaign_charity[:giving_sg_url])
+    end
+
+    @campaign.campaign_charities = campaign_charities
+  end
+
+  def final_scope
+    scope = scoped_with_name
+
+    scope = scope.merge(scoped_with_dates)
+    return scope if start_params? || end_params?
+
+    status_params = params[:status]
+    return scope if status_params.nil?
+
+    scope.merge(scoped_with_status)
+  end
+
+  def scoped_with_name
+    params[:name].present? ? Campaign.contains(params[:name]) : Campaign.all
+  end
+
+  def start_params?
+    start_params = params[:start]
+    start_params.present? && start_params['from'].present? && start_params['to'].present?
+  end
+
+  def end_params?
+    end_params = params[:end]
+    end_params.present? && end_params['from'].present? && end_params['to'].present?
+  end
+
+  def scoped_with_dates
+    start_params = params[:start]
+    starts_before = start_params['to'] if start_params.present?
+    starts_after = start_params['from'] if start_params.present?
```

Comment:
Doesn't start_params return a boolean itself?

Reply:
That's `start_params?` though, this is `start_params`.